packages/components/src/components/files/FileTree.tsx

```diff
@@ -0,0 +1,386 @@
+                    }) as List<string>,
+                }),
+                () => this.onFileSelect(node.id, checked, !!node.children)
+            );
+        }
+    };
+
+    // recursively toggle all child files. afterCascade used to check selection box of each subfile
+    cascadeToggle = (node, afterCascade: (any) => any): void => {
+        const afterToggle = () => {
+            afterCascade(node);
+            if (node.children) {
+                node.children.forEach(child => {
+                    this.cascadeToggle(child, afterCascade);
+                });
+            }
+        };
+
+        this.onToggle(node, true, afterToggle);
+    };
+
+    handleCheckbox = (evt): void => {
+        const { data } = this.state;
+        const { checked } = evt.target;
+        const id = this.getNodeIdFromId(evt.target.id);
+
+        const node = this.getDataNode(id, data);
+        if (node.children && checked) {
+            // Toggle open selected directory and check the children
+            const callback = checkNode => {
+                this.setCheckedValue(checkNode, checked);
+            };
+
+            this.cascadeToggle(node, callback);
+        }
+        this.setCheckedValue(node, checked);
+    };
+
+    loadDirectory = (node: any, callback?: () => any): any => {
+        const { loadData } = this.props;
+
+        this.setState(() => ({loading: true}));
+        loadData(this.getPathFromId(node.id))
+            .then(children => {
+
+                const { data } = this.state;
+                let dataNode = this.getDataNode(node.id, data);
+
+                children = children.map(child => {
+                    child.id = dataNode.id + '|' + child.name; // generate Id from path
+                    return child;
+                });
+
+                if (children.length < 1) {
+                    children = [{ id: dataNode.id + '|' + this.EMPTY_FILE_NAME, active: false, name: 'empty' }];
+                }
+
+                dataNode.children = children; // This is not immutable so this is updating the data object
+                this.setState(
+                    () => ({ cursor: node, data: Object.assign({}, data), error: undefined, loading: false }),
+                    callback
+                );
+            })
+            .catch((reason: any) => {
+                this.setState(() => ({ error: reason.message ? reason.message : 'Unable to fetch data', loading: false }));
+            });
+    };
+
+    onToggle = (node: any, toggled: boolean, callback?: () => any): void => {
+        const { cursor, data } = this.state;
+
+        if (cursor) {
+            node.active = false;
+        }
+        node.active = true;
+        node.toggled = toggled;
+
+        if (node.children) {
+            // load data if not already loaded
+            if (node.children.length === 0) {
+                node.children = [{ id: node.id + '|' + this.LOADING_FILE_NAME }];
+                this.setState(
```

Comment:
It also appears that in this method `node` comes from `this.state`, and we are directly modifying the object, which the react docs say you should not do.

Reply:
Object.assign changed to spread.

Yeah the node passed back from treebeard is a pointer into this.state.data (which is a tree of files and directories).  I could right off the bat make a copy of this.state.data and do a search through the data tree for the node to operate on but this seems like an unnecessary traversal of the file tree looking for the node when there's already a pointer to the node on the existing this.state.data tree.  Seems more efficient to just operate on the node/data in state and then do the clone of data for setState.  I know its not the way we want to operate on state normally, but not seeing a big danger in this case.

In that setState, the order of state change needs to be, directory node goes to loading state, then directory node loads data.  Putting loadDirectory into the callback just ensures it happens in that order.  I changed the parameter on loadDirectory from node to just node.id so the callback will have its own copy of the primitive.  Otherwise updating state in loadDirectory doesn't really matter when it happens as long as the state change to loading comes first.